### PR TITLE
mac80211: enable ATH9K_HWRNG

### PR DESCRIPTION
**Pull request by neheb**
in kernel 5.17, fcd09c90c3c5254b18ef34e30c57c65d34290a84 integrated it better with thee random framework.

Gives boot time randomness on supported devices.

### CODE REVIEW

File: package/kernel/mac80211/ath.mk [hunk 1]
```diff
@@ -377,7 +377,7 @@ endef
 define KernelPackage/ar5523
   $(call KernelPackage/mac80211/Default)
   TITLE:=Driver for Atheros AR5523 USB sticks
-  DEPENDS:=@USB_SUPPORT +kmod-mac80211 +kmod-ath +kmod-usb-core +kmod-input-core
```
Comment (BKPepe) [inline package/kernel/mac80211/ath.mk:380]: This is unrelated change to your commit

File: package/kernel/mac80211/ath.mk [hunk 2]
```diff
@@ -225,7 +225,7 @@ define KernelPackage/ath9k/config
 		bool "Add wireless noise as source of randomness to kernel entropy pool"
 		depends on PACKAGE_kmod-ath9k
 		select PACKAGE_kmod-random-core
-		default n
+		default y
```
Comment (BKPepe) [inline package/kernel/mac80211/ath.mk:228]: I think OpenWrt core members would be interested in how much does it make a difference in storage capacity.

File: package/kernel/mac80211/ath.mk (hunk 1 above)
Comment (neheb) [reply package/kernel/mac80211/ath.mk:380]: My text editor automatically does this

**Comment by neheb** [comment]
ERROR: package/utils/busybox failed to build (build variant: default).

can't compile a build.

**Comment by neheb** [comment]
ok this is hilarious. Enabling this option results in a size reduction.

Before:
-rw-r--r--. 1 mangix mangix  5463422 Mar  4 23:00 openwrt-ath79-generic-tplink_archer-c7-v2-initramfs-kernel.bin

After:
-rw-r--r--. 1 mangix mangix  5463264 Mar  4 23:03 openwrt-ath79-generic-tplink_archer-c7-v2-initramfs-kernel.bin

I have no explanation for this.

**Comment by DragonBluep** [comment]
> ok this is hilarious. Enabling this option results in a size reduction.

Lzma compression ratio is affected by many factors. The size change should be negligible as `kmod-random-core` is already a dependency package of the ath9k.

The problem is whether it can really work. I have an AR9287 device. The noise showed by `iwinfo` is always a fixed value -95 dBm. Maybe some old ath9k devices do not support this feature? Need ath9k developer to confirm.
```
phy0-ap0  ESSID: "OpenWrt"
          Access Point: 00:40:5A
          Mode: Master  Channel: 1 (2.412 GHz)  HT Mode: HT40
          Center Channel 1: 3 2: unknown
          Tx-Power: 23 dBm  Link Quality: 15/70
          Signal: 116 dBm  Noise: -95 dBm
          Bit Rate: 300.0 MBit/s
          Encryption: none
          Type: nl80211  HW Mode(s): 802.11b/g/n
          Hardware: 168C:002E 168C:A199 [Atheros AR9287]
          TX power offset: none
          Frequency offset: none
          Supports VAPs: yes  PHY name: phy0
```

**Comment by neheb** [comment]
yeah after a full wipe I now get 5462678. no change in sysupgrade size. Anyway.

I don't think the noise in iwinfo and the hwrng are related. There was a bunch of testing done when this was merged. Older devices already don't work with this and will not initialize. See https://github.com/rsalvaterra/openwrt/commit/692eefad9367b43c9e07778a8dae3dea117ca84b

**Comment by DragonBluep** [comment]
Confirm that random numbers can be generated on ar9287.

```
root@OpenWrt:~#  xxd -l 256 /dev/urandom
00000000: ab35 d16d 327a 042c 17e3 0b75 dcc1 e91c  .5.m2z.,...u....
00000010: 60ff b461 7daf 5967 963b b85b 6679 779f  `..a}.Yg.;.[fyw.
00000020: 029c b581 41b1 c379 e29c feae 0e71 8484  ....A..y.....q..
00000030: f88a 8f17 fc14 9b8e f8d1 3a6e 03d4 b39e  ..........:n....
00000040: 99fa ffd8 fa36 8c60 7043 20f7 174b c7d9  .....6.`pC ..K..
00000050: 87c6 8113 d40d e869 bbee 9a5e 6137 0a65  .......i...^a7.e
00000060: 9e45 1a8b 85d6 a0e1 ee98 924a b555 c3e9  .E.........J.U..
00000070: 2b1a 15b6 ea75 b0de 21f4 23e4 9831 2359  +....u..!.#..1#Y
00000080: 44b0 3e0f 4631 786e 0a9d b1d4 ceeb 75c8  D.>.F1xn......u.
00000090: c097 c693 fc22 e762 a123 4fa0 c20e 4195  .....".b.#O...A.
000000a0: 4d5e 4dec 6249 59f5 3d33 b0c7 be3a f708  M^M.bIY.=3...:..
000000b0: d834 31b7 5607 a6f4 c9bf 2f38 5fdc a581  .41.V...../8_...
000000c0: 6af6 0cfb bb1f 5d1a 5273 5d94 f7bb 85f1  j.....].Rs].....
000000d0: 3435 21f6 2325 dde6 d7cc 24fb 73d4 50ed  45!.#%....$.s.P.
000000e0: 4c6d a6a2 c9f5 33ce f894 f998 00c5 a2ff  Lm....3.........
000000f0: 9b99 4ff6 7e7d 4aef 72a3 4842 f38e 69e6  ..O.~}J.r.HB..i.
root@OpenWrt:~#  xxd -l 256 /dev/urandom
00000000: 51bd 7594 85c3 616c fa0e e14a 77df c173  Q.u...al...Jw..s
00000010: a6ca b596 6519 ccc9 ab25 67ec e1ea 1fb1  ....e....%g.....
00000020: 415b 4911 5485 393a 3fbe ecc4 ddc9 562c  A[I.T.9:?.....V,
00000030: 4ac0 6324 40d1 0935 8cb0 6610 9489 0875  J.c$@..5..f....u
00000040: 2e28 389b d910 0575 ec6a d966 a0f0 2215  .(8....u.j.f..".
00000050: 8a59 dbe5 8edf b292 c15f 7b5e e929 7253  .Y......._{^.)rS
00000060: fcdf 180d 1276 0740 3867 242f 799f 7cfe  .....v.@8g$/y.|.
00000070: 6243 fcc0 6836 d661 4d29 4d68 8f4b 3b3e  bC..h6.aM)Mh.K;>
00000080: 0189 138b 7c3d 34a0 b220 3cb8 4136 2ff0  ....|=4.. <.A6/.
00000090: 2bad 3612 5bea 837e c90f 194e 983d 6385  +.6.[..~...N.=c.
000000a0: 480b bc4d d72d 1ca5 c5f3 fa11 a813 ad47  H..M.-.........G
000000b0: afdc b9c3 4483 4494 7462 eb14 db28 d09b  ....D.D.tb...(..
000000c0: cf73 c80f 9aa3 e50a b030 5826 55c2 700d  .s.......0X&U.p.
000000d0: bb9d 3ba9 1d7a 6617 5a37 8210 a5a0 90f9  ..;..zf.Z7......
000000e0: af75 1fa2 31e2 87a7 9719 d8f1 3278 a554  .u..1.......2x.T
000000f0: 27bb 9257 3c2f aadc cb83 54c6 321a 33f7  '..W</....T.2.3.

root@OpenWrt:~# xxd -l 256 /dev/random
00000000: 485a c445 d337 1973 07de bf1e 644c e0e9  HZ.E.7.s....dL..
00000010: 5d6f d9e3 bcb5 8aa5 ee66 e30d bee1 37cb  ]o.......f....7.
00000020: 03e5 3799 4d6c c1de 729f 125d e5df 1c18  ..7.Ml..r..]....
00000030: 090a a4f2 6ddf 452a e606 1a17 d535 dbad  ....m.E*.....5..
00000040: 2955 dc65 7650 5247 5583 5294 2ecd eaaf  )U.evPRGU.R.....
00000050: 5090 7d5b 8cda f35b 7621 3b6b f754 6990  P.}[...[v!;k.Ti.
00000060: a4f0 b25f bffc 2dd9 5363 a9a4 d395 a03f  ..._..-.Sc.....?
00000070: 066e 3d7f 10b3 a892 9f28 8784 d0df 0433  .n=......(.....3
00000080: 68aa d07c 40dd 029a b26f ee5a 9c56 4f1d  h..|@....o.Z.VO.
00000090: 9ebb eae8 3e7d 29ce 7a9a 8cac 82cf eaee  ....>}).z.......
000000a0: baaf 6875 4620 82bb bd0a b35e ac82 a90b  ..huF .....^....
000000b0: 1d08 bb8d 2d0e f4cc 0a8b 7f2c b197 a531  ....-......,...1
000000c0: b591 c1e2 346c 69a8 e41a 3843 e7b9 8c3d  ....4li...8C...=
000000d0: bc2a 4b4e cccf ec24 e899 65a5 c560 8022  .*KN...$..e..`."
000000e0: aa24 132e f5b6 3076 b44b c831 b75c 76e2  .$....0v.K.1.\v.
000000f0: 1794 1400 926f 8d25 ce87 2628 fc17 1eaa  .....o.%..&(....
root@OpenWrt:~# xxd -l 256 /dev/random
00000000: a51b 5335 8754 dfe3 a31a 79a2 0f4c 87fc  ..S5.T....y..L..
00000010: 0f82 e9be 34bc 8ca0 78ae a8c6 4264 3de7  ....4...x...Bd=.
00000020: 4c22 b0a3 a6cb 4718 91e9 ba16 cd91 4acc  L"....G.......J.
00000030: 9390 5c0b 0526 c954 66a2 c02d 8a25 2b50  ..\..&.Tf..-.%+P
00000040: 8b33 3705 a610 8340 bf22 f254 8003 8817  .37....@.".T....
00000050: 1e9a 85a0 391c af77 602d 0214 41cb 36d1  ....9..w`-..A.6.
00000060: 09d5 df60 143c 37de 2a24 cbc9 ba35 bdda  ...`.<7.*$...5..
00000070: 58eb 32ec 4c32 a180 c983 f605 5cf0 552f  X.2.L2......\.U/
00000080: f462 9f3d 801f 7a1d 1748 3849 9528 efb1  .b.=..z..H8I.(..
00000090: eba6 5418 b624 fb42 6208 e6ae 4f3f 812a  ..T..$.Bb...O?.*
000000a0: 2b13 e04a 6ee5 b14a 9b38 7384 d987 17de  +..Jn..J.8s.....
000000b0: 8571 503a 7987 82d5 308f a88a ca55 910c  .qP:y...0....U..
000000c0: 7270 18a7 bfc7 a176 af73 1654 079e c230  rp.....v.s.T...0
000000d0: 2d06 99a5 df20 734d c57b 7cf6 380a 4025  -.... sM.{|.8.@%
000000e0: 31c2 de7c 2df0 43b1 2797 be40 311b ec6c  1..|-.C.'..@1..l
000000f0: 6f03 0064 cd6c 1ece f376 c230 7a66 8c14  o..d.l...v.0zf..

```

**Comment by hauke** [comment]
Thanks for looking at the sizes.

The hardware random number generator does not use the noise level, but it should use the noise itself. The noise it receives when not connected to a Wifi should be pretty random.

Between the HW RNG and /dev/random is still a pseudo random number generator. Even when the ath9k device would generates only 0, /dev/random would still look random.
